### ISSUE TITLE
UnifiedFFI: FFIExternalStructureReferenceHandle is not polymorphic with ExternalAddress

### ISSUE THREAD

**Issue by theseion** [guillep/pharo-core#4452]
`FFIExternalStructureReferenceHandle` is used to for accessing struct fields. So when I use a method for accessing a field I will get such a handle. Unfortunately, I can't use that handle in place of other handles (usually `ExternalAddress`) as `#getHandle` answers the start of the struct, not of the field.

I can work around this issue by sending `#asExternalAddress` to the structure reference handle but that's inconvenient and I don't think that uses should need to know about the difference between the two.

**Comment by Ducasse** [comment]
Hi max
What G&P told me is that the FFI part for accessing structures have to be fully redone. I do not know if this is related to this issue but probably. 
S.

**Comment by theseion** [comment]
Yeah, that's what I heard. Just wanted to record this, so it won't be forgotten.

**Comment by theseion** [comment]
@guillep @tesonep any updates on this? Can I close it?

**Comment by guillep** [comment]
Sorry I never checked this.
So let me rephrase, what you want is to get an external address that has the computed address of a field (base+offset)?
Could you sketch an example of what you want (ideally something we could turn into a test)?

**Comment by theseion** [comment]
It's been a long time so I had to go look at code. As far as I recall, the issue came up while generating field accessors. Take this example from `LGitBlameHunk>>prim_final_commit_id`:

```smalltalk
prim_final_commit_id
	"This method was automatically generated"
	^ LGitId fromHandle: (handle referenceStructAt: OFFSET_PRIM_FINAL_COMMIT_ID length: LGitId byteSize) asExternalAddress
```

The `asExternalAddress` there is not part of the generated code but a manual addition. The generator correctly generates the rest of the code but the handle accepted by LGitId must be an `ExternalAddress`, not a `FFIExternalStructureReferenceHandle`. Or rather, it could be a `FFIExternalStructureReferenceHandle` but that class is not polymorphic with `ExternalAddress`.

The weird thing is that the generator should actually take care of this, because `LGitId` is just a normal `FFIExternalStructure`. Take, for example, `FFIStructure>>pointer`:

```smalltalk
pointer
	self getHandle isExternalAddress 
		ifFalse: [ self error: 'Structures need to be moved to external memory space before passing them as pointers.' ].
	^ self getHandle pointer
```

Presuming that the structure was initialized above (without `#asExternalAddress), `#getHandle` here would return an instance of `FFIExternalStructureReferenceHandle`. `FFIExternalStructureReferenceHandle` implements `isExternalAddress` as:

```smalltalk
isExternalAddress
	^ handle isExternalAddress
```

This would answer true (the handle of the reference is an ExternalAddress). But then `^ self getHandle pointer` will fail with an error because `FFIExternalStructureReferenceHandle` *does not* override `Object>>pointer` (which simply signals an error). This is just an example to show that I don't understand how it could ever work to pass a `FFIExternalStructureReferenceHandle` as the handle of a structure instead of an `ExternalAddress`.

**Comment by tinchodias** [comment]
Hello, I ran into this issue [when adding support for a Harfbuzz API in Alexandrie](https://github.com/pharo-graphics/Alexandrie/commit/ec77e357ab8c053ce2573cf6264aba34a30ead09#diff-7067f54908da59dc9185a365c1c016cca233d539be87a8fb7eac9a612257a6f9R57).

[The CI log: shows the issue](https://github.com/pharo-graphics/Alexandrie/actions/runs/7061922041/job/19224612055)
```
AeHbFeatureTest
 ✗ #testParseInstanceInArray (1ms)
Instance of FFIExternalStructureReferenceHandle did not understand #tfPointerAddress
FFIExternalStructureReferenceHandle(Object)>>doesNotUnderstand: #tfPointerAddress
AeHbFeature(ExternalObject)>>tfPointerAddress
AeHbFeature>>readIntoSelf:size:
AeHbFeature>>readIntoSelf:
AeHbFeatureTest>>testParseInstanceInArray ...readIntoSelf: '+kern[1:7]'
AeHbFeatureTest(TestCase)>>performTest
```

In my local image it works when I define the method as follows, in `FFIExternalStructureReferenceHandle`:
```
tfPointerAddress
	^ handle tfPointerAddress + startOffset
```
I placed it in the protocol "*ThreadedFFI-UFFI" and will create a pull request.

I wrote it based on the following method in the same class:
```
asExternalAddress
	^ handle + startOffset
```
(BTW it's an extension of Freetype package)

**Comment by tinchodias** [comment]
I test it in a workspace with this script:

```smalltalk 
anArray := FFIExternalArray newType: FFITestStructure size: 3.
addresses := anArray collect: #tfPointerAddress as: Array.
"an Array(@ 16r100045F3B50 @ 16r100045F3B88 @ 16r100045F3BC0)"

(addresses second - addresses first) asInteger = FFITestStructure byteSize. "true"
```

Am I missing something?

**Comment by tinchodias** [comment]
Well, I wasn't exactly talking of making `FFIExternalStructureReferenceHandle` polymorphic with `ExternalAddress`.
But I think #15666 is related
